### PR TITLE
[projmgr,cbuildgen] support options (optimize, debug, warnings)

### PR DESCRIPTION
**Pull request by tarek-bochkati**
Contributed by STMicroelectronics
Signed-off-by: Tarek BOCHKATI <tarek.bouchkati@st.com>

### CODE REVIEW

**Comment by tarek-bochkati** [comment]
@brondani we still have to propagate this to CMakeLists.txt and manage it in <TOOLCHAIN>.cmake files

File: tools/projmgr/src/ProjMgrGenerator.cpp [hunk 1]
```diff
@@ -219,6 +219,21 @@ void ProjMgrGenerator::GenerateCprjVector(XMLTreeElement* element, const vector<
...
+    return;
+  }
+
+  XMLTreeElement *optionsElement = element->CreateElement("options");
+  if (optionsElement)
+  {
+    SetAttribute(optionsElement, "optimize", buildType.optimize);
+    SetAttribute(optionsElement, "debug", buildType.debug);
+    SetAttribute(optionsElement, "warnings", buildType.warnings);
+  }
```
Comment (brondani) [inline tools/projmgr/src/ProjMgrGenerator.cpp:234]: @tarek-bochkati Codecov fails due to decreased code coverage in this PR. Could you please add a test case to cover lines 228-234? Thanks!

**Review by brondani — COMMENTED** [review]
@tarek-bochkati Thanks for this contribution! Please see my comment concerning the code coverage results.

File: tools/projmgr/src/ProjMgrGenerator.cpp (hunk 1 above)
Comment (tarek-bochkati) [reply tools/projmgr/src/ProjMgrGenerator.cpp:234]: @brondani I have added a test that I expect it to cover this code chunk.

File: tools/projmgr/src/ProjMgrGenerator.cpp [hunk 2]
```diff
@@ -167,7 +167,7 @@ void ProjMgrGenerator::GenerateCprjTarget(XMLTreeElement* element, const Context
     targetOutputElement->AddAttribute("outdir", context.directories.outdir);
   }
 
-  // TODO Generate toolchain settings (warnings, debug, includes)
+  GenerateCprjOptions(element, context.cproject->target.build);
```
Comment (tarek-bochkati) [inline tools/projmgr/src/ProjMgrGenerator.cpp:170]: @brondani I'm wondering if this is the correct build type container ?
I have been thinking about the case when there is options like "optimize" in the csolution and not in the cproject
IMO this should be propagated here as well.
even more, if optimize is specified in both csolution (optimize: size) and cproject(optimize: speed), then I think "speed" should override "size" optimization

Comment (tarek-bochkati) [reply tools/projmgr/src/ProjMgrGenerator.cpp:170]: then for the time being let's keep this as is without overriding
as you suggested, let's move this discussion to #179

**Comment by tarek-bochkati** [comment]
@brondani I have added more commits to propagate the options to CMakeLists.txt
now I working on supporting these options on IAR.cmake

File: tools/buildmgr/cbuild/src/CbuildModel.cpp [hunk 3]
```diff
@@ -1048,23 +1067,67 @@ const bool CbuildModel::EvalFlags() {
...
   }
 
   return true;
 }
 
-const bool CbuildModel::EvalItemTranslationControls(const RteItem* item, bool isFlag, const string& groupName) {
+const bool CbuildModel::EvalOptions() {
+  /*
+  EvalFlags:
+  Evaluate Flags
```
Comment (brondani) [inline tools/buildmgr/cbuild/src/CbuildModel.cpp:1081]: Missing update after copy&paste?

**Comment by tarek-bochkati** [comment]
@brondani there was a bug on `tools/projmgr/schemas/common.schema.json` related to `WarningsType` 
I will come back with another fix for codecov

**Comment by jkrech** [comment]
@tarek-bochkati  can you please ensure that the open conversations above get closed commenting how these comments have been addressed or if case they have not been addressed, why this is.

File: tools/projmgr/src/ProjMgrGenerator.cpp (hunk 2 above)
Comment (tarek-bochkati) [reply tools/projmgr/src/ProjMgrGenerator.cpp:170]: I added a comment there as suggested

File: tools/buildmgr/cbuildgen/src/CMakeListsGenerator.cpp [hunk 4]
```diff
@@ -40,10 +40,13 @@ bool CMakeListsGenerator::GenBuildCMakeLists(void) {
   if (!m_targetSecure.empty())       cmakelists << EOL << "set(SECURE " << m_targetSecure << ")";
   if (!m_targetMve.empty())          cmakelists << EOL << "set(MVE " << m_targetMve << ")";
   if (!m_byteOrder.empty())          cmakelists << EOL << "set(BYTE_ORDER " << m_byteOrder << ")";
-  if (!m_asMscGlobal.empty())       cmakelists << EOL << "set(AS_FLAGS_GLOBAL \"" << CbuildUtils::EscapeQuotes(m_asMscGlobal) << "\")";
-  if (!m_ccMscGlobal.empty())       cmakelists << EOL << "set(CC_FLAGS_GLOBAL \"" << CbuildUtils::EscapeQuotes(m_ccMscGlobal) << "\")";
-  if (!m_cxxMscGlobal.empty())      cmakelists << EOL << "set(CXX_FLAGS_GLOBAL \"" << CbuildUtils::EscapeQuotes(m_cxxMscGlobal) << "\")";
-  if (!m_linkerMscGlobal.empty())   cmakelists << EOL << "set(LD_FLAGS_GLOBAL \"" <<CbuildUtils::EscapeQuotes( m_linkerMscGlobal) << "\")";
+  if (!m_optimize.empty())           cmakelists << EOL << "set(OPTIMIZE " << m_optimize << ")";
+  if (!m_debug.empty())              cmakelists << EOL << "set(DEBUG " << m_debug << ")";
```
[the remark this answers is not in the document]
Comment (tarek-bochkati) [reply tools/buildmgr/cbuildgen/src/CMakeListsGenerator.cpp]: I have changed all options names as following OPT_DEBUG, OPT_WARNINGS, OPT_OPTIMIZE

File: tools/projmgr/test/data/TestSolution/test.target_options.cproject.yml [hunk 5]
```diff
@@ -0,0 +1,8 @@
+# yaml-language-server: $schema=https://raw.githubusercontent.com/Open-CMSIS-Pack/devtools/schemas/projmgr/0.9.0/tools/projmgr/schemas/cproject.schema.json
+
+project:
+  optimize: size
+  debug: on
```
[the remark this answers is not in the document]
Comment (tarek-bochkati) [reply tools/projmgr/test/data/TestSolution/test.target_options.cproject.yml]: done, please check the new `tools/projmgr/test/data/TestSolution/test.target_options.cproject.yml`

**Comment by tarek-bochkati** [comment]
@brondani could you please give some hints on how to fix the buildmgr-cov ?

**Comment by brondani** [comment]
> @brondani could you please give some hints on how to fix the buildmgr-cov ?

@tarek-bochkati in the [codecov report](https://app.codecov.io/gh/Open-CMSIS-Pack/devtools/compare/328/diff) you can see the new lines that are not being covered by tests and probably causing the coverage to decrease. For example [lines 895 - 904 in CbuildModel.cpp](https://github.com/Open-CMSIS-Pack/devtools/blob/a8a5ea48a589dedae0f9a043d4017892f08767d4/tools/buildmgr/cbuild/src/CbuildModel.cpp#L895-L904) are shown as not being covered by tests.

File: tools/buildmgr/cbuildgen/src/BuildSystemGenerator.cpp [hunk 6]
```diff
@@ -98,6 +98,10 @@ bool BuildSystemGenerator::Collect(const string& inputFile, const CbuildModel *m
   m_asMscGlobal = GetString(model->GetTargetAsFlags());
   m_linkerMscGlobal = GetString(model->GetTargetLdFlags());
 
+  m_optimize = model->GetTargetOptimize();
+  m_debug = model->GetTargetDebug();
+  m_warnings = model->GetTargetWarnings();
+
```
Comment (brondani) [inline tools/buildmgr/cbuildgen/src/BuildSystemGenerator.cpp:104]: The options need to be handled also at component/group/file level in addition to target level.

Comment (tarek-bochkati) [reply tools/buildmgr/cbuildgen/src/BuildSystemGenerator.cpp:104]: Well spotted, but before proceed I prefer to wait for:
- concluding on https://github.com/Open-CMSIS-Pack/devtools/pull/360 and https://github.com/Open-CMSIS-Pack/devtools/issues/179
- and I need to understand how `*flags` are managed at component/group/file level in [cbuildgen]

**Comment by tarek-bochkati** [comment]
@brondani is it possible to run codecov locally, so I can detect coverage regressions before pushing to github ?

**Comment by brondani** [comment]
> @brondani is it possible to run codecov locally, so I can detect coverage regressions before pushing to github ?

Please take a look here:
https://github.com/Open-CMSIS-Pack/devtools/blob/main/.github/workflows/buildmgr.yml#L478-L493
@soumeh01 Could you please add to the devtools documentation how to generate the coverage report locally? Thanks!

**Comment by tarek-bochkati** [comment]
as mentioned by @brondani , this change is too huge
decision is to split it into multiple PRs : we started with #478

**Comment by tarek-bochkati** [comment]
summary:
- #478 : merged
- #479 : merged
- #480 : merged
- #484 : under review

**Pull request closed without merging**